### ISSUE TITLE
consider implementing "VER" to display the $GIT_STRING

### ISSUE THREAD

**Issue by Ben-401** [MEGA65/mega-basic64#2]

**Comment by lgblgblgb** [comment]
Hmm, and/or let's display the version number on the start-up screen, as with the C65, in various ROMs. Though, currently M65-basic seems to use the C64-style startup screen in this design, so maybe it wouldn't fit so nicely ...

**Comment by gardners** [comment]
Hello,

I'll have a think about how we might make this accessible.

Paul.

**Comment by Ben-401** [comment]
hi, i would like to be able for the PRG program to know what version of 
bitstream it is using, not so much for the end-user. Yes as said the 
user can see from the boot screen (if they are quick).

I understand there is already a function in ASM that returns the O/S 
version or something in reg .X and .Y, but i was thinking in BASIC if 
VER were typed, that the GITstring would be displayed to the screen, or 
even captured and parsed as a variable.

this come about because when using the monitor_load program, the 
boot-screen does not appear visible long enough to read it. Yes the 
M65DBG could be used, but i already have a different terminal program 
connected to the uart.

thanks Paul.

**Comment by gardners** [comment]
Hello,

If the goal is to see the boot messages, just hold the CONTROL key down
when resetting, and it will stay visible until you let go.

Paul.

**Comment by bobbyjim** [comment]
It sounds like he's interested in writing code that's version-aware.

**Comment by gardners** [comment]
Correct, but then it should be into a variable, eg, ver$, not display to
the screen, else it would be a pain to access.

**Comment by bobbyjim** [comment]
Sounds good!  That is totally in line with how Commodore BASIC works.

**Comment by lgblgblgb** [comment]
One problem with the variable, that AFAIK, CBM (well MS ...) BASIC only uses the first two letters to distinguish between variables, even if the name is longer. So a variable like VER$ is actually only VE$. I always find that problematic to introduce new special variables because of this, it can easily break compatibility with existing programs using _any_ variable name starting with "VE" (even if it's just VE$ and not $VER ... or let it be VE2$ or anything). So maybe a "function" is better idea, but then a new BASIC token must be introduced just because of this :( I already saw problems with DS which is disk status or such in later CBM BASICs, breaking compatibility with existing BASIC programs (DS was not "reserved variable" in earlier CBM BASICs ...).

**Comment by bobbyjim** [comment]
I. YET ANOTHER VARIABLE
Perhaps the easiest solution; however, polluting the two-character variable namespace is bound to cause collisions (and subsequent indigestion). 

**Ia. EXPAND THE VALID VARIABLE CHARACTER INVENTORY WITH _ or @ or ...**
Perl stole ideas from BASIC, so steal ideas from Perl.  Use one of the currently-unusable character combinations to store environmental data.  The back arrow, plus $, or up arrow plus $.  Underscore, plus $.  Maybe even $$.  As long as BASIC can figure out what's a variable and what isn't, then something like this could work.  Ugly?  Can be, but not necessarily ($$ is not ugly.  Even _0$ through _9$ is not too bad, or up-arrow-0$ through up-arrow-9$).

I note that @ is not a current variable ID for strings, but it's right next to A.  So how about adding it to the allowed characters for variable IDs, then assigning environment variables using it?

**I really like this one idea.**

II. AN ENVIRONMENTAL FUNCTION
For example, ENV$(x).  A function (and another token), where the index parameter is leveraged to return all sorts of information.  For example, something like 0 for $GIT_STRING.  1 for CPU uptime.  2 for free memory (never a negative number).  

Allows incremental scope creep.

IIa. An Environmental function that accepts a string parameter.  The readability is probably not worth the cost.

III. CRAZIER WHACK-JOB BRAINSTORMS
A. LONG VARIABLE NAMES
Make it THREE characters, or eight, or sixteen, or 256.    Then add environmental variables and give them readable names.

GIT$
UPTIME
FREEMEM

B. AN ENVIRONMENT VARIABLE NAMESPACE
Introduce a separate, read-only string namespace.  Probably not worth the effort.

GIT_STRING@
UPTIME@
FREEMEM@
